File: src/main/java/gauge/execution/runner/GaugeRerunFailedAction.java
package gauge.execution.runner;

import com.intellij.execution.Executor;
import com.intellij.execution.configurations.GeneralCommandLine;
import com.intellij.execution.configurations.RunProfile;
import com.intellij.execution.configurations.RunProfileState;
import com.intellij.execution.runners.ExecutionEnvironment;
import com.intellij.execution.testframework.actions.AbstractRerunFailedTestsAction;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.ui.ComponentContainer;
import com.intellij.psi.search.GlobalSearchScope;
import gauge.GaugeConstants;
import gauge.execution.GaugeCommandLine;
import gauge.execution.GaugeCommandLineState;
import gauge.execution.GaugeRunConfiguration;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class GaugeRerunFailedAction extends AbstractRerunFailedTestsAction {
  public GaugeRerunFailedAction(@NotNull ComponentContainer componentContainer) {
    super(componentContainer);
  }

  @Nullable
  @Override
  protected MyRunProfile getRunProfile(@NotNull ExecutionEnvironment environment) {
    RunProfile config = myConsoleProperties.getConfiguration();
    return config instanceof GaugeRunConfiguration ? new RerunProfile((GaugeRunConfiguration)config) : null;
  }

  private static class RerunProfile extends MyRunProfile {

    private final GaugeRunConfiguration config;

    RerunProfile(GaugeRunConfiguration configuration) {
      super(configuration);
      this.config = configuration;
    }

    @Nullable
    @Override
    public RunProfileState getState(@NotNull Executor executor, @NotNull ExecutionEnvironment env) {
      GeneralCommandLine commandLine = GaugeCommandLine.getInstance(config.getModule(), getProject());
      commandLine.addParameters(GaugeConstants.RUN, GaugeConstants.FAILED);
      return new GaugeCommandLineState(commandLine, getProject(), env, config);
    }

    @Nullable
    @Override
    public GlobalSearchScope getSearchScope() {
      return null;
    }

    @Override
    public void checkConfiguration() {
    }

    @Override
    public Module @NotNull [] getModules() {
      return this.config.getModules();
    }
  }
}

File: src/main/kotlin/gauge/execution/runner/GaugeRerunFailedAction.kt
package gauge.execution.runner

import com.intellij.execution.Executor
import com.intellij.execution.configurations.RunProfileState
import com.intellij.execution.runners.ExecutionEnvironment
import com.intellij.execution.testframework.actions.AbstractRerunFailedTestsAction
import com.intellij.openapi.module.Module
import com.intellij.openapi.ui.ComponentContainer
import com.intellij.psi.search.GlobalSearchScope
import gauge.GaugeConstants
import gauge.execution.GaugeCommandLine
import gauge.execution.GaugeCommandLineState
import gauge.execution.GaugeRunConfiguration

class GaugeRerunFailedAction(componentContainer: ComponentContainer) :
    AbstractRerunFailedTestsAction(componentContainer) {

    override fun getRunProfile(environment: ExecutionEnvironment): MyRunProfile? {
        val config = myConsoleProperties.configuration
        return if (config is GaugeRunConfiguration) RerunProfile(config) else null
    }

    private class RerunProfile(private val config: GaugeRunConfiguration) : MyRunProfile(config) {

        override fun getState(executor: Executor, env: ExecutionEnvironment): RunProfileState {
            val commandLine = GaugeCommandLine.getInstance(config.selectedModule, project)
            commandLine.addParameters(GaugeConstants.RUN, GaugeConstants.FAILED)
            return GaugeCommandLineState(commandLine, project, env, config)
        }

        override fun getSearchScope(): GlobalSearchScope? {
            return null
        }

        override fun checkConfiguration() {
            // No-op
        }

        override fun getModules(): Array<Module> {
            return config.modules
        }
    }
}
